chains/evm/contracts/pools/USDC/USDCTokenPoolProxy.sol

```diff
@@ -0,0 +1,372 @@
+    }
+
+    // In previous versions of the USDC Token Pool, the sourcePoolData only contained two fields, a uint64 and uint32.
+    // For structs stored only in memory, the compiler assigns each field to its own 32-byte slot, instead of tightly
+    // packing like in storage. This means that a message originating from a previous version of the pool will have a
+    // sourcePoolData that is 64 bytes long, indicating an inflight message originating from a previous version of
+    // the USDC Token pool.
+    // This branch must come before a version check, because the first field would be a uint64 and thus if a version
+    // was attempted to be extracted from the first 4-bytes of a uint64, it would be 0, and thus the message would be
+    // routed to the CCTP V1 pool without first sanitizing the source pool data for proper formatting.
+    if (releaseOrMintIn.sourcePoolData.length == 64) {
+      // There are two possible scenarios for the legacy inflight messages:
+      // 1. The legacy pool did not utilize a message transmitter proxy.
+      // 2. The legacy pool utilized a message transmitter proxy, but the format of the sourcePoolTokenData was as described
+      // in the comments above.
+
+      // In the first scenario, only the message's destinationCaller, I.E the legacy pool, can execute the mint, and so
+      // the message needs to be routed to the legacy pool. In the second scenario, the destinationCaller will be the
+      // message transmitter proxy, and the message needs to be routed to the appropriate V1-compatible pool.
+      if (_checkForLegacyInflightMessages(releaseOrMintIn.offchainTokenData)) {
+        // Note: Supporting this branch will require this proxy to be set as an offRamp in the router, which is a design
+        // decision that is not ideal, but allows for a direct upgrade from the first version of the USDC Token Pool to
+        // this version.
+        return USDCTokenPool(s_pools.legacyCctpV1Pool).releaseOrMint(releaseOrMintIn);
+      } else {
+        // Since the new pool and the inflight message should utilize the same version of CCTP, and would have the same
+        // destinationCaller (the message transmitter proxy), we can route the message to the v1 pool, but we first
+        // need to turn the source pool data into the new format, otherwise the decoding scheme will fail. Once there is
+        // confidence that no more messages are inflight, these branches can be safely removed.
+
+        // Since the CCTP v1 pool will have this contract set as an allowed caller, no additional configurations are
+        // needed to route the message to the v1 pool.
+        return USDCTokenPool(s_pools.cctpV1Pool).releaseOrMint(_generateNewReleaseOrMintIn(releaseOrMintIn));
+      }
+    }
+
+    // The first 4 bytes of source pool data are the version which can be extracted directly and cast into a uint32.
+    bytes4 version = bytes4(releaseOrMintIn.sourcePoolData[:4]);
```

Comment:
Can we now safely move this up to be above the length check? I kinda worry about length checks causing false positives. If we do tag checks first that should be solved